server/conn.go

```diff
@@ -1295,6 +1317,80 @@ func (cc *clientConn) handleQuery(ctx context.Context, sql string) (err error) {
 	return err
 }
 
+// prefetchPointPlanKeys extracts the point keys in multi-statement query,
+// use BatchGet to get the keys, so the values will be cached in the snapshot cache, save RPC call cost.
+func (cc *clientConn) prefetchPointPlanKeys(stmts []ast.StmtNode) ([]plannercore.Plan, error) {
+	txn, err := cc.ctx.Txn(false)
+	if err != nil {
+		return nil, err
+	}
+	if !txn.Valid() {
+		// Only build in-transaction query for simplicity.
+		// Later we can support out-transaction multi-statement query.
+		return nil, nil
+	}
+	vars := cc.ctx.GetSessionVars()
+	if vars.TxnCtx.IsPessimistic && vars.TxnCtx.Isolation == ast.ReadCommitted {
```

Comment:
maybe it has a limited effect when `IsPessimistic == true && vars.TxnCtx.Isolation == RR`？ because it's still need lock per stmt(and lock request also can prefetch cache), but it's great for multi-stmt in side optimistic txn :D

Reply:
Yes, I should make it skip pessimistic transactions.